Sources/Bond/ObservableCollection/ObservableCollection.swift

```diff
@@ -0,0 +1,94 @@
+//
+//  The MIT License (MIT)
+//
+//  Copyright (c) 2018 Tony Arnold (@tonyarnold)
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Differ
+import ReactiveKit
+
+public class ObservableCollection<UnderlyingCollection: Collection>: SignalProtocol {
+
+    public internal(set) var collection: UnderlyingCollection
+    internal let subject = PublishSubject<ObservableCollectionEvent<UnderlyingCollection>, NoError>()
```

Comment:
Is there any way this could be public? I have a custom collection conformance that isn't an array, and I'd like to be able to feed events through to it manually myself.

Reply:
I also have times where I need to send an event that's already occurred. Creating new instances in this custom collection time automatically inserts them into the collection, which means I need to send the event without ever performing an insert.

Reply:
Oh! Can I use `descriptiveUpdate(…)` for this? Nice!

Reply:
Yeah, ignore all of that - `descriptiveUpdate` does _exactly_ what I need. Man, you're a smart cookie!

Reply:
Yeah, I wanted to hide the subject as an implementation detail. `descriptiveUpdate` should be used to update the collection and provide diff if known. Will that work for you?

Reply:
Alright, great! :)

Reply:
`NSTableView` code works 👌 

Nicely done - this is awesome, and has solved so many problems for me - I've been able to remove an entire layer of abstraction due to this WIP PR!

Reply:
> I'm missing some details regarding move operations, but other operations should work

Ah yes, I just hit this, I think. All good, this is looking fantastic! Thank you so much for your work.

Reply:
These should now be implemented - would be great if you could try and see if you are still hitting some errors :)

Reply:
I think there might be something awry in the move handling. 

If I have:

```
1, 2, 3, 4
```

and I ask to move `1` to `4`, then `2` to `4`, I get:

```
1, 3, 4, 2
```

When I would have expected I would get:

```
1, 4, 2, 3
```

Reply:
I'm seeing some unusual behaviour around batched moves. I'll try to get some meaningful data together about it, but the actual moves being produced appear to be incorrect.

Reply:
Alright, it would be great if you could find an example that is wrong.

Reply:
OK, so it looks like there's something busted with batch move patches: 

I have an array:

```
[1, 2, 3, 4]
```

I'm going to perform the following action: `fromIndexes: [2, 3], toIndex: 1`, which produces a diff that looks like this:

```
[M(2, 1), M(3, 1)]
```

The resulting patch looks like:

```
[M(2, 1), M(2, 1)]
```

Which when followed through step by step:

1. `[1, 3, 2, 4]`
2. `[1, 2, 3, 4]`

I'll have a look at the code now, but that's the one that's tripping me up.

Reply:
That patch looks good to me. I don't understand your action though, you seem to want to move two elements to the same index.

Reply:
I have a `move(fromIndexes: IndexSet, to toIndex: Int)` method - it removes the elements from their original indexes, and re-inserts them in order at the specified `toIndex`.

Reply:
Ah, ok. I think it boils down to how to apply such event.  Give me a minute.

Reply:
I tried to construct the moves manually using `descriptiveUpdate {}` and applying each move in sequence, wrapped in a `batchUpdate {}` without much luck.

Reply:
I think you want to product the following diff:

```
M(from[0], to+0)
M(from[1], to+1)
...
```

And then in `descriptiveUpdate` update the collection (without using `batchUpdate`) and then send that diff.

Reply:
I might be easier to update the collection by doing delete + insert. You would send diff of moves, of course :)

Reply:
The actual change to the data is fine - I have a method on my collection that handles it (and I've verified as working) - it's the construction of the `CollectionOperation.move` that's tripping me up. 

I'll keep playing and bring you something concrete if I find it. Thanks @srdanrasic!

Reply:
Are the operations returned from `descriptiveUpdate()` used as-is, or are they converted to patches for things like `NSTableView`?

Reply:
So the diff you send from `descriptiveUpdate` is used as is. You (`NSTableView`) can then call `.patch` on it to get a sequence of patch events (implemented in CollectionOperation+Patch.swift).

Reply:
@tonyarnold this is what I had in mind

```swift
extension MutableObservableCollection where UnderlyingCollection: RangeReplaceableCollection, UnderlyingCollection.Index: Strideable, UnderlyingCollection.Index.Stride == Int {

    public func moveItems(from fromIndexes: [UnderlyingCollection.Index], to toIndex: UnderlyingCollection.Index) {
        descriptiveUpdate { (collection) -> [CollectionOperation<UnderlyingCollection.Index>] in
            let items = fromIndexes.map { collection[$0] }
            for index in fromIndexes.sorted().reversed() {
                collection.remove(at: index)
            }
            collection.insert(contentsOf: items, at: toIndex)
            return fromIndexes.enumerated().map {
                CollectionOperation<UnderlyingCollection.Index>.move(from: $0.element, to: toIndex.advanced(by: $0.offset))
            }
        }
    }
}
```

Reply:
Thanks for looking into this, @srdanrasic. I really appreciate it.

So that still doesn't seem to do what I would like - I plugged the extension into the Table View Binding playground:

```swift
let c = MutableObservableCollection(["1", "2", "3", "4"])
c.moveItems(from: [2,3], to: 1)
```

That code generates a diff of:

```
[M(2, 1), M(3, 2)]
```

Which is resulting in a patch of:

```
[M(2, 1), M(2, 2)]
```

That patch seems suspect to me. The second move does nothing to the tableview when it's applied (so I end up with the data in the right state `["1", "3", "4", "2"]`, but the NSTableView is showing `["1", "3", "2", "4"]`).

This seems like it should be simple enough to solve, but my brain is doing loops 🤕

Reply:
I think the patch is fine from the point of view that it represents a sequence of changes that when applied to the original array one after another give you the final array. 

The question is - is that what NSTableView expects. From the documentation it seems yes, but I guess no.  

Btw, is it possible to have NSTableView in a playgrounds as live view? I couldn’t get it to show.

Reply:
I’m still uncertain if NSTableView wants diff or patch when changes are wrapped in begin/endUpdates.

Reply:
Even within a batch, I believe NSTableView needs patches. The underlying indexes are updated immediately after the call to insert/remove/move - it’s just the view updates that are held back.

Reply:
It should be possible to show a table view in a playground, but I’m not near my Mac so I can’t check, sorry!

Reply:
Alright, I’ll be working on this PR tomorrow so I’ll check out NSTableView too. 

Have a nice day away from the Mac! 🌞

Reply:
Thanks! I'll sit down tonight and get you some answers about the playground support - but yes, time away from the computer is 🌻

Reply:
OK, I couldn't get NSTableView working in a playground, so here's a sample project that implements the absolute minimum for drag-rearranging an `ObservableCollection`: 
[Table Move Test.zip](https://github.com/DeclarativeHub/Bond/files/1910260/Table.Move.Test.zip)

Reply:
Thanks for that, I see now where the problem is! I'll try to fix it today.

Reply:
@tonyarnold I made some improvements to the patch algorithm so it should now properly handle moves. I might have missed some corner cases so I plan to write some unit tests there. 

I've also detected an issue with batchUpdate when there is a specific combination of operations inside - will work on that tomorrow. 

`moveItem` is renamed to `move`. `move(from fromIndices: [Index], to toIndex: Index)` is now part of the framework. 

Your test app should now work, at least the dropping part. Note that you have to manually calculate `to` index:

```
    func tableView(_ tableView: NSTableView, acceptDrop info: NSDraggingInfo, row: Int, dropOperation: NSTableView.DropOperation) -> Bool {
        // Unpack the source indexes from the drag
        let indexData = info.draggingPasteboard().data(forType: .stringData)

        guard
            let _indexes = indexData.flatMap(NSKeyedUnarchiver.unarchiveObject(with:)) as? IndexSet,
            _indexes.isEmpty == false
        else {
            return false
        }

        let indexes = _indexes.map { $0 }
        strings.moveItems(from: indexes, to: row - indexes.filter { $0 < row}.count)

        return true
    }
```

I haven't tested the dragging part. Not sure how to trigger that 😄

Reply:
Fantastic, thanks @srdanrasic! I can’t wait to have a look at this (when the kids leave me be for 30 minutes!)

Reply:
> I haven't tested the dragging part. Not sure how to trigger that

I'm not sure what you mean? You should just be able to drag with your mouse - select a row or two and drag away!

This all seems to work nicely. Have you considered placing the row offset logic inside the move method, like so?

```swift
public func move(from fromIndices: [UnderlyingCollection.Index], to toIndex: UnderlyingCollection.Index) {
    descriptiveUpdate { (collection) -> [CollectionOperation<UnderlyingCollection.Index>] in
        let adjustedToIndex = toIndex - fromIndices.filter { $0 < toIndex }.count
        collection.move(from: fromIndices, to: adjustedToIndex)
        return fromIndices.enumerated().map {
            .move(from: $0.element, to: adjustedToIndex.advanced(by: $0.offset))
        }
    }
}
```

Then you could just pass the raw index from `NSTableView`?

Reply:
I might have broke something regarding dragging then. 

Batch updates should now work. I've completely refactored the internal algorithm that merges diffs. I feel I got it right this time - it super exhausting to think about how these indices are shifting back and forth 😅 

As for `move(from fromIndices: [UnderlyingCollection.Index], to toIndex: UnderlyingCollection.Index)`, I think that the current implementation makes sense. `to` refers to the staring index that the elements will be moved to. It's NSTableView that is giving us a different kind of an index - one that refers to the source array instead of the final one. 

If we would make the adjustment in our method, we would be specializing the method only for NSTableView usage.

Reply:
> we would be specializing the method only for NSTableView usage

Yeah, fair call. 

Thank you so much for looking into this when I needed it - I owe you. I think this PR has made a pretty great change to Bond - it's certainly freed things up to support what I needed. I'll try to get the observable tree work tidied up this coming week, too.